### PR TITLE
Add workflow translation id to classification metadata

### PR DESCRIPTION
**Pull request by mcbouslog**
Staging branch URL: https://pr-5127.pfe-preview.zooniverse.org/
- helpful test project: https://pr-5127.pfe-preview.zooniverse.org/projects/eatyourgreens/-project-testing-ground

Towards #5125.

- adds translation resource id to corresponding translations store strings type
- saves `workflow_translation_id` and `user_language` to classification metadata on completion
- renames `updateClassification` to `updateMetadata` to better reflect actual action

Questions
- [ ] **would this record the necessary info so we could search classifications and check for mismatched workflow translations?**
- [x] preferred default `null`? could be empty string or something else. some of my default/initial settings might not be necessary.

Apologies for the excessive commits, if merging please squash.

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
Should the translation ID be added to metadata or links? All the other resource IDs are recorded in `classification.links`?

What's your thinking behind adding a new action creator, rather than adding the translation ID as an extra argument to either classification creation or classification completion? My thinking would have been to add it to the creation function, along with the project, workflow and subject but I could be missing something.

File: app/pages/project/classify.jsx [hunk 1]
```diff
@@ -82,7 +86,11 @@ export class ProjectClassifyPage extends React.Component {
   }
 
   componentDidUpdate(prevProps) {
-    const { classification, upcomingSubjects, workflow } = this.props;
+    const { actions, classification, translations, upcomingSubjects, workflow } = this.props;
+
+    if (translations !== prevProps.translations) {
```
Comment (eatyourgreens) [inline app/pages/project/classify.jsx]: `translations` is a nested object, so I'm not sure that you can use direct comparison here to detect changes.

File: app/redux/ducks/classify.js [hunk 2]
```diff
@@ -32,11 +32,12 @@ function awaitSubjectSet(workflow) {
   }
 }
 
-function createNewClassification(project, workflow, subject, goldStandardMode) {
+function createNewClassification(project, workflow, subject, goldStandardMode, workflowTranslationId) {
   const source = subject.metadata.intervention ? 'sugar' : 'api';
   const classification = apiClient.type('classifications').create({
     annotations: [],
     metadata: {
+      workflow_translation_id: workflowTranslationId,
```
Comment (eatyourgreens) [inline app/redux/ducks/classify.js]: Is this here because it won't work in the links object?

**Comment by eatyourgreens** [comment]
I'm seeing the workflow translation action fire before the classification is created, which I think is a bug.
Redux dev tools showing the order in which the classifier actions fire.

If I change the project language from English to Italian, I see the workflow translation action fire multiple times, with a null payload.
Redux dev tools showing the workflow translation action firing multiple times on language change.

**Comment by eatyourgreens** [comment]
I'm seeing the workflow translation action fire before the classification is created, which I think is a bug.
Redux dev tools showing the order in which the classifier actions fire.

If I change the project language from English to Italian, I see the workflow translation action fire multiple times, with a null payload.
Redux dev tools showing the workflow translation action firing multiple times on language change.

**Comment by eatyourgreens** [comment]
I'm seeing the workflow translation action fire before the classification is created, which I think is a bug.
Redux dev tools showing the order in which the classifier actions fire.

If I change the project language from English to Italian, I see the workflow translation action fire multiple times, with a null payload.
Redux dev tools showing the workflow translation action firing multiple times on language change.

**Comment by eatyourgreens** [comment]
It would help if I used a workflow that actually has an Italian version, right?

If I change the language, the translation ID changes but I don't see the classification change. Here, the classification is recording the English language translation.
Redux dev tools showing classifier state after a language change.

**Comment by eatyourgreens** [comment]
Since I can change language during the course of classifying, it probably makes more sense to record the language and translation ID at the end of the classification, not the start.

**Comment by eatyourgreens** [comment]
With the latest commit, do we still need the workflow translation action?

**Comment by mcbouslog** [comment]
I think the action can be removed if saving on completion.

This is likely overkill, not necessary, but do we want to save the workflow translation id at creation and then separately at completion? I don't think so, but thought I'd ask.

**Comment by mcbouslog** [comment]
I think the action can be removed if saving on completion.

This is likely overkill, not necessary, but do we want to save the workflow translation id at creation and then separately at completion? I don't think so, but thought I'd ask.

**Comment by mcbouslog** [comment]
Originally went the link route, but per a brief Slack conversation with @srallen and @camallen we thought we'd go with saving it in the metadata.

Once this is in good shape it might be worth having @marten or @camallen confirm this will provide the necessary info so we can search all classifications for mismatches.

**Comment by eatyourgreens** [comment]
I think quite a lot of code can go,  in favour of using `actions.classify.updateClassification()` to change classification metadata directly from `props.translations`.

**Comment by eatyourgreens** [comment]
Incidentally, I noticed while working on the classifier that we only ever update classification metadata, which is why that function doesn't update any other classification properties.

**Comment by mcbouslog** [comment]
I'll remove some code and update to store an array of translation ids.

**Comment by eatyourgreens** [comment]
Maybe that `updateClassification` function would be better named `updateMetdata`? That would be consistent with the naming of the `updateAnnotations` function.

File: app/pages/project/classify.jsx (hunk 1 above)
Comment (srallen) [reply app/pages/project/classify.jsx]: lodash has a deep equality function: https://lodash.com/docs/4.17.11#isEqual

Comment (eatyourgreens) [reply app/pages/project/classify.jsx]: In other components, I've listened for locale changes in order to detect when the language changes.
https://github.com/zooniverse/Panoptes-Front-End/blob/eb1a1556d22405adea855fbac1bd6c58d9ec2fac/app/pages/project/workflow-selection.jsx#L39-L41

Comment (eatyourgreens) [reply app/pages/project/classify.jsx]: Also note that you could dispatch an `updateClassification` action from within the translations ducks file, whenever a new workflow translation loads. Interventions does this in order to update the classifier store from Sugar messages:
https://github.com/zooniverse/Panoptes-Front-End/blob/eb1a1556d22405adea855fbac1bd6c58d9ec2fac/app/redux/ducks/interventions.js#L3

**Comment by mcbouslog** [comment]
Ok I think this is looking better now. I like the array of workflow translation ids, captures user behavior better.

I'm indifferent between the current approach where on classifier translations prop change it updates the metadata vs updating the metadata from within the translations duck. I kind of like keeping the translations duck without importing other actions or checking the translation type then doing something unique, but I don't feel strongly about that at all, not really strong technical arguments.

**Comment by mcbouslog** [comment]
While a classification metadata property `workflow_translation_ids` as an array of null or translation resource ids captures user behavior well, I worry it might make searching classifications for mismatched workflows and translations significantly harder or impossible?

If so, and if the immediate goal is to track if the workflow-translation bug is still occuring, we might want to just save the `initial_workflow_translation_id` as a string to metadata? As I think that initial classifier load with the mismatch is what we want to capture in metadata.

Roping in @camallen or @marten for their thoughts on this point.

**Comment by eatyourgreens** [comment]
I’m not sure about capturing an array of translations either, since we only record the user language once, at the end of the classification.

**Comment by camallen** [comment]
>While a classification metadata property workflow_translation_ids as an array of null or translation resource ids captures user behavior well, I worry it might make searching classifications for mismatched workflows and translations significantly harder or impossible?
>If so, and if the immediate goal is to track if the workflow-translation bug is still occuring, we might want to just save the initial_workflow_translation_id as a string to metadata? As I think that initial classifier load with the mismatch is what we want to capture in metadata.

Is it an array of translation ids to capture the different language strings? Correct me where i'm not getting it. 

As it stand in code now, I assume for say an default language project (let's say english from the lab) it'll record just eng language translation id. If however i change languages, do we want to record this change and track which strings were shown? 

If so then instead of an array of ids, I think i'd prefer the metadata object to be an object keyed by language code for review in exports, etc `{ 'en-au': 12345, 'en-gb': 12354 }` i think that'll help get a sense of what the user saw without having to lookup the translation record manually (and help us track the state of the classifier as it changes, thus what the user actually saw).

If you just want to track the translation id for the submitted language strings, then why does it need to be in an array? just plonk it into a `{ 'workflow_translation_id': 12345 }`  attribute in the metadata object.

**Comment by mcbouslog** [comment]
Per Slack conversation we've decided to save the workflow translation id on completion.

No projects identified as possibly affected by the workflow-translation bug have multiple translations. Therefore, we believe it's not necessary to save multiple translation ids to the classification metadata.

File: app/pages/project/classify.jsx [hunk 3]
```diff
@@ -182,7 +184,7 @@ export class ProjectClassifyPage extends React.Component {
   loadAnotherSubject() {
     const { actions, project, workflow } = this.props;
     if (workflow) {
-      actions.classifier.nextSubject(project, workflow);
+      actions.classifier.nextSubject(project);
```
Comment (eatyourgreens) [inline app/pages/project/classify.jsx]: Out of interest, why was the workflow removed here?

Comment (mcbouslog) [reply app/pages/project/classify.jsx]: Per https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/redux/ducks/classify.js#L315-L320 I don't think the workflow param is used? The classify duck tracks workflow in it's state (https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/redux/ducks/classify.js#L164)?

Comment (mcbouslog) [reply app/pages/project/classify.jsx]: This also applies to line 123.

File: app/classifier/classifier.jsx [hunk 4]
```diff
@@ -253,15 +253,17 @@ class Classifier extends React.Component {
...
         width: innerWidth,
         height: innerHeight
       },
       interventions: {
         message: !!showIntervention,
         opt_in: !!user && user.intervention_notifications
-      }
+      },
+      workflow_translation_id: translations.strings.workflow.id,
+      user_language: translations.locale
```
Comment (eatyourgreens) [inline app/classifier/classifier.jsx]: Small point, but I think I forgot to add tests for this change.

Comment (mcbouslog) [reply app/classifier/classifier.jsx]: Just added.

File: app/redux/ducks/translations.js [hunk 5]
```diff
@@ -40,7 +40,7 @@ const initialState = {
   rtl: false,
   strings: {
     project: {},
-    workflow: {},
+    workflow: { id: null },
```
Comment (wgranger) [inline app/redux/ducks/translations.js]: Why was this `id: null` added?

Comment (mcbouslog) [reply app/redux/ducks/translations.js]: If there's no translation for the workflow I wanted to provide a default or null value so when on classification completion `translations.strings.workflow.id` is saved we'll know there wasn't a translation, as opposed to just not saving anything and wondering if it worked at all. So I added that default id of null in the duck as well as the classifier component.

Comment (eatyourgreens) [reply app/redux/ducks/translations.js]: Does the API client record a value if you pass `null`? I thought it removed the key in that case. I could be wrong though.

**Comment by eatyourgreens** [comment]
Tests are passing. 👍

File: app/redux/ducks/translations.js (hunk 5 above)
Comment (wgranger) [reply app/redux/ducks/translations.js]: Looks like if no id is present, `workflow_translation_id: null` is recorded in the metadata.

Comment (mcbouslog) [reply app/redux/ducks/translations.js]: I just did a few classifications on a testing project without any workflow translations and ran a data export and I think we're ok:
screen shot 2018-12-07 at 11 00 45 am

Comment (eatyourgreens) [reply app/redux/ducks/translations.js]: Cool. Sounds like we didn't roll out translations on staging though. All projects should have them at this point, even if it's just the one English translation that you get by default.

**Pull request merged**